onnxruntime/core/providers/cuda/math/clip.cc

```diff
@@ -35,11 +47,38 @@ Status Clip<T>::ComputeInternal(OpKernelContext* ctx) const {
   return Status::OK();
 }
 
-#define SPECIALIZED_COMPUTE(T) \
-  REGISTER_KERNEL_TYPED(T)     \
-  template Status Clip<T>::ComputeInternal(OpKernelContext* ctx) const;
+template <typename T>
+Status Clip<T>::ComputeInternal(OpKernelContext* ctx) const {
+  const auto* min = ctx->Input<Tensor>(1);
+  const auto* max = ctx->Input<Tensor>(2);
+
+  auto min_val = -std::numeric_limits<T>::infinity();
+  auto max_val = std::numeric_limits<T>::infinity();
+  if (min) {
+    ORT_ENFORCE(min->Shape().NumDimensions() == 0, "min should be a scalar.");
+    min_val = *(min->template Data<T>());
+  }
+  if (max) {
+    ORT_ENFORCE(max->Shape().NumDimensions() == 0, "max should be a scalar.");
+    max_val = *(max->template Data<T>());
+  }
+  ORT_ENFORCE(min_val <= max_val);
+
+  const Tensor& X = *ctx->Input<Tensor>(0);
+  const TensorShape input_shape{X.Shape()};
+  Tensor* Y = ctx->Output(0, input_shape);
+  size_t count = input_shape.Size();
+
+  if (count > 0) {
+    auto* y_data = Y->template MutableData<T>();
+    const auto* x_data = X.template Data<T>();
+    ClipImpl<T>(x_data, y_data, min_val, max_val, count);
+  }
+
+  return Status::OK();
```

Comment:
many duplicate code with the existing one. is it possible judge the version from how many inputs you can get?

Reply:
you can also get the start and end version like this : info.GetKernelDef().SinceVersion(&start, &end);

Reply:
Good suggestion. Will update.